### PR TITLE
Feat: aws-sdk instrumentation

### PR DESCRIPTION
**Pull request by YanivD**
Related to issue https://github.com/open-telemetry/opentelemetry-ruby/issues/253

During the development I looked on 2 examples:
* [aws-xray instrumentation](https://github.com/aws/aws-xray-sdk-ruby/blob/master/lib/aws-xray-sdk/facets/aws_sdk.rb)
* [datadog instrumentation](https://github.com/DataDog/dd-trace-rb/tree/master/lib/ddtrace/contrib/aws)

This is my first PR in Ruby, so please review this PR carefully 🙏

### CODE REVIEW

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (arielvalentin) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:25]: ✨ I really like this! 

@open-telemetry/ruby-maintainers I do realize we don't love using `untraced` but I think this would be great to include as an option for other instrumentation where we may have redundant client instrumentation (if we are not doing so already)

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb
Comment (arielvalentin) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb:53]: Consider adding some diagnostic output here using the `OpenTelemetry.error_handler` or `OpenTelemetry.logger` to hint that a specific constant could not be loaded.

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/services.rb
Comment (arielvalentin) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/services.rb:7]: 😅 This quite a library!

**Review by arielvalentin — APPROVED** [review]
This looks good to me! Thank you very much for this excellent contribution. 

I will defer to any other reviewers who have more experience than I using the AWS instrumentation.

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (arielvalentin) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:59]: Question for AWS users... is there anyway to look up the position/priority of an existing Handler like we can do with Rails middleware?

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (YanivD) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:25]: @arielvalentin 
Agree!
It reduce the amount of redundant spans and making the trace much clearer.

Many nodejs otel instrumentations accept this option, like [opentelemetry-instrumentation-aws-sdk](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/node/opentelemetry-instrumentation-aws-sdk/README.md#aws-sdk-instrumentation-options).
`suppress_internal_instrumentation` option should always be `false` by default to make sure adding new instrumentations won't remove existing spans.

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb
Comment (YanivD) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb]: Good idea. Added in [this commit](https://github.com/open-telemetry/opentelemetry-ruby/pull/1014/commits/432902cd6a45ec0b3054baa2eff0f428b059952d)

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (mwear) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:37]: The RPC semantic conventions comply with the specification: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-sdk.md, however, the `aws.region` semantic convention isn't formally specified. OTel JS [uses it](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-aws-sdk#span-attributes) though.

I am wondering if the Resource semantic convention of setting `cloud.provider` and `cloud.region` would be a better option (see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/cloud.md#cloud). Happy to have others chime in on the best approach here.

**Review by mwear — COMMENTED** [review]
I just have a couple of comments / questions, but otherwise this looks really good. Thanks for the contribution @YanivD!

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (YanivD) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb]: Good idea, splitted to 2 methods

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (YanivD) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb]: I used [xray as inspiration](https://github.com/aws/aws-xray-sdk-ruby/blob/master/lib/aws-xray-sdk/facets/aws_sdk.rb#L21-L22) for service name extraction.
Added safe navigation in [this commit](https://github.com/open-telemetry/opentelemetry-ruby/pull/1014/commits/31ed27fa58aa9a42a45fc1289409caac506e87d6#diff-1723a57ddb33bc724a6da2be5c6a3dffa0a655cb65a43fbe00a78c71f5d2cb89R44).

---

I tried to use safe navigation for every method in the chain, but I got error from `rubocop`. 
@mwear What is the best practice in this case?

### What I tried:
```ruby
def get_service_name(context)
  context&.client&.class&.api&.metadata['serviceId'] || context&.client&.class&.to_s&.split('::')[1]
end
```

### The warning I get:
> lib/opentelemetry/instrumentation/aws_sdk/handler.rb:44:48: W: Lint/SafeNavigationChain: Do not chain ordinary method call after safe navigation operator.
>           context&.client&.class&.api&.metadata['serviceId'] || context&.client&.class&.to_s&.split('::')[1]

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (YanivD) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:36]: You're raising a good question. 

I'm not sure resource attributes are a good fit for this attribute. 
As I understand, resource attributes are immutable representation of the running service.
For example, if I'm running a service on EC2, `cloud.provider` and `cloud.region` resource attributes should keep the value of the running environment. If the running service is producing `sqs.send_message` operation to a queue in a different region, `cloud.region` should have the EC2's region and not the SQS destination region.

Maybe we should use `cloud.destination.region` instead of `aws.region`.

The cloud semantic conventions are still in Experimental status. This topic is not covered yet. 
IMO we can keep the behavior as implemented in OTEL JS, and update it in the future when it will have a clear definition.

WDYT?

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (ericmustin) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:14]: The specification requires that we set `db.system: dynamodb` in the event that this is a request for dynamodb. I think this would be possible to determine based on the context 

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-sdk.md#dynamodb

Could we add support+a test for this as it's required (if there are optional attributes you think are worth adding from dynamoDB, the more the merrier, but this is the only one that's required as per the specification)?

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-sdk.md#dynamodb

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb
Comment (ericmustin) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb:16]:
```suggestion
          require_relative 'handler'
          require_relative 'services'
```

**Review by ericmustin — CHANGES_REQUESTED** [review]
A few small nits, generally lgtm nice work!

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (YanivD) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:14]: @ericmustin Nice catch! 
[Added `db.system` attribute](
https://github.com/open-telemetry/opentelemetry-ruby/pull/1014/commits/cfd81abada4f85dbdbaaa893729b668ec8f7e198)

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (ericmustin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb]: chaining after `.to_s` is unnecessary in this case, and likely  `.class` as well. basically if the class exists then it would be reasonable to assume `.class` would return as that method exists on every class, and likewise with `.to_s`. Does that help?

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (ericmustin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:36]: my 2 cents is, unless we want to put in the effort to make an otep or spec change upstream , stick with what js does. i think `cloud.destination.region` is a completely reasonable choice fwiw, but I just think practically speaking lets lean on the experimental status here and let ppl speak up if they hate the attr name.

**Comment by ericmustin** [comment]
Lgtm pending some of the feedback from @mwear getting addressed/resolved. nice work here ty for the contribution!

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (YanivD) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb]: Thanks @ericmustin!

Can you confirm this chaining is safe?
```rb
context&.client.class.api.metadata['serviceId'] || context&.client.class.to_s.split('::')[1]
```

Comment (ericmustin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb]: I have no idea off hand, i'd have to look into the implementation 🙃. I can follow up when i have some more bandwidth. 

specifically, are `client`, `api` and `metadata` always guaranteed to return what you expect here? If it's possible they return `nil` then we'd want to chain those with the safe operator

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (mwear) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:36]: i'm also fine with sticking `aws.region` as is for now.

**Review by mwear — APPROVED** [review]
LGTM. Thanks for addressing my feedback!

File: instrumentation/aws_sdk/test/opentelemetry/instrumentation_test.rb
Comment (arielvalentin) [inline instrumentation/aws_sdk/test/opentelemetry/instrumentation_test.rb:57]: 💁🏼 I think what we want here is [`must_be_nil`](https://docs.seattlerb.org/minitest/Minitest/Expectations.html#method-i-must_be_nil)
```suggestion
       _(last_span.attributes['db.system']).must_be_nil
```

File: instrumentation/aws_sdk/test/opentelemetry/instrumentation_test.rb
Comment (arielvalentin) [inline instrumentation/aws_sdk/test/opentelemetry/instrumentation_test.rb:72]:
```suggestion
       _(last_span.attributes['db.system']).must_be_nil
```

File: instrumentation/aws_sdk/test/opentelemetry/instrumentation_test.rb
Comment (arielvalentin) [inline instrumentation/aws_sdk/test/opentelemetry/instrumentation_test.rb:86]:
```suggestion
         _(last_span.attributes['db.system']).must_be_nil
```

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (arielvalentin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb]: @ericmustin I was referring to this. In the case context is `nil` it will always results in a `NoMethodError` because `NilClass` does not have a class method `api`. 

```
omoplata ~(:|✔) %
🤘 irb
irb(main):001:0> context = nil
=> nil
irb(main):002:0> context&.client.api
Traceback (most recent call last):
        4: from /Users/arielvalentin/.rbenv/versions/2.7.4/bin/irb:23:in `<main>'
        3: from /Users/arielvalentin/.rbenv/versions/2.7.4/bin/irb:23:in `load'
        2: from /Users/arielvalentin/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        1: from (irb):2
NoMethodError (undefined method `api' for nil:NilClass)
```

Comment (ericmustin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb]: ah far as i can tell, `context`, `client`, and `api` should be initialized with default values and never return `nil`

- `context` https://github.com/aws/aws-sdk-ruby/blob/b5c46409986c6681b35533e017051d55e733d053/gems/aws-sdk-core/lib/seahorse/client/response.rb#L13

- `client` https://github.com/aws/aws-sdk-ruby/blob/b5c46409986c6681b35533e017051d55e733d053/gems/aws-sdk-core/lib/seahorse/client/request_context.rb#L39-L40

- `api` https://github.com/aws/aws-sdk-ruby/blob/b5c46409986c6681b35533e017051d55e733d053/gems/aws-sdk-core/lib/seahorse/client/base.rb#L171-L173

- `metadata` https://github.com/aws/aws-sdk-ruby/blob/b5c46409986c6681b35533e017051d55e733d053/gems/aws-sdk-core/lib/seahorse/model/api.rb#L8

Technically I guess someone could provide a junk value for context, but many other parts of the aws-sdk would break in that situation anyway, so i'm not too concerned.

tl;dr, i think this is much ado about nothing, 

`context.client.class.api.metadata['serviceId'] || context.client.class.to_s.split('::')[1]` is fine imo

**Review by ericmustin — APPROVED** [review]

**Comment by YanivD** [comment]
Is there anything else missing before merge and release?

@open-telemetry/ruby-maintainers

**Comment by mwear** [comment]
I believe everything has been addressed. Any objections to me merging this @robertlaurin and @fbogsany?

**Review by robertlaurin — APPROVED** [review]

**Pull request merged**
